File: lab3/msLab3/src/classes/Element.ts
import { DistributionEnum, RoutingEnum } from "../utils/enums.js";
import { Route } from "./Route.js";
import { Job } from "./Job.js";
import * as random from "../utils/randomizers.js";

export class Element {
  private static nextId = 0;
  private routes: Route[] = [];
  private id: number;
  private name: string;
  private routing: RoutingEnum = RoutingEnum.BY_PRIORITY;
  private distribution: DistributionEnum;
  private tNext: number;
  private tCurr: number;
  private delayMean: number;
  private delayDev: number;
  private quantity: number = 0;
  private state: number = 0;

  constructor(name: string, delayMean: number = 1.0, delayDev?: number) {
    this.name = name;
    this.tNext = delayDev === undefined ? Infinity : 0.0;
    this.tCurr = this.tNext;
    this.delayMean = delayMean;
    this.delayDev = delayDev || 0;
    this.distribution =
      delayDev === undefined ? DistributionEnum.NONE : DistributionEnum.NORMAL;
    this.id = Element.nextId++;
  }

  private static getUnblockedRoutes(routes: Route[], routedJob: Job): Route[] {
    return routes.filter((route) => !route.isBlocked(routedJob));
  }

  private static getScaledProbabilities(routes: Route[]): number[] {
    const probabilities = routes.map(
      (route, i) =>
        route.getProbability() + (i === 0 ? 0 : probabilities[i - 1])
    );
    const totalProbability = probabilities[probabilities.length - 1];
    return probabilities.map((prob) => prob / totalProbability);
  }

  public setRouting(routing: RoutingEnum): void {
    this.routing = routing;
  }

  public getDelay(): number {
    switch (this.distribution) {
      case DistributionEnum.EXPONENTIAL:
        return random.randomExponential(this.delayMean);
      case DistributionEnum.UNIFORM:
        return random.randomUniform(this.delayMean, this.delayDev);
      case DistributionEnum.NORMAL:
        return random.randomNormal(this.delayMean, this.delayDev);
      case DistributionEnum.ERLANG:
        return random.randomErlang(this.delayMean, this.delayDev);
      default:
        return this.delayMean;
    }
  }

  public getDelayMean(): number {
    return this.delayMean;
  }

  public setDelayMean(delayMean: number): void {
    this.delayMean = delayMean;
  }

  public getDelayDev(): number {
    return this.delayDev;
  }

  public setDelayDev(delayDev: number): void {
    this.delayDev = delayDev;
  }

  public getQuantity(): number {
    return this.quantity;
  }

  public setQuantity(quantity: number): void {
    this.quantity = quantity;
  }

  public changeQuantity(delta: number): void {
    this.quantity += delta;
  }

  public getNextRoute(routedJob: Job): Route {
    if (this.routes.length === 0) {
      return new Route(null);
    }

    switch (this.routing) {
      case RoutingEnum.BY_PROBABILITY:
        return this.getNextRouteByProbability(routedJob);
      case RoutingEnum.BY_PRIORITY:
        return this.getNextRouteByPriority(routedJob);
      case RoutingEnum.COMBINED:
        return this.getNextRouteCombined(routedJob);
    }
  }

  private getNextRouteByProbability(routedJob: Job): Route {
    const unblockedRoutes = Element.getUnblockedRoutes(this.routes, routedJob);
    if (unblockedRoutes.length === 0) {
      return this.routes[0];
    }

    const probability = Math.random();
    const scaledProbabilities = Element.getScaledProbabilities(unblockedRoutes);

    for (let i = 0; i < scaledProbabilities.length; i++) {
      if (probability < scaledProbabilities[i]) {
        return unblockedRoutes[i];
      }
    }

    return unblockedRoutes[unblockedRoutes.length - 1];
  }

  private getNextRouteByPriority(routedJob: Job): Route {
    const unblockedRoutes = Element.getUnblockedRoutes(this.routes, routedJob);
    return unblockedRoutes.length === 0 ? this.routes[0] : unblockedRoutes[0];
  }

  private getNextRouteCombined(routedJob: Job): Route {
    let selectedRoute: Route | null = null;

    for (const route of this.routes) {
      if (!route.isBlocked(routedJob)) {
        selectedRoute = route;
        break;
      }
    }

    if (!selectedRoute) {
      return this.routes[0];
    }

    const samePriorityRoutes = this.findRoutesByPriority(
      selectedRoute.getPriority()
    );
    const probability = Math.random();
    const scaledProbabilities =
      Element.getScaledProbabilities(samePriorityRoutes);

    for (let i = 0; i < scaledProbabilities.length; i++) {
      if (probability < scaledProbabilities[i]) {
        return samePriorityRoutes[i];
      }
    }

    return selectedRoute;
  }

  private findRoutesByPriority(priority: number): Route[] {
    return this.routes.filter((route) => route.getPriority() === priority);
  }

  public addRoutes(...routes: Route[]): void {
    this.routes.push(...routes);
    this.routes.sort((a, b) => b.getPriority() - a.getPriority());
  }

  public inAct(job: Job): void {}

  public outAct(): void {
    this.quantity++;
  }

  public getTNext(): number {
    return this.tNext;
  }

  public setTNext(tNext: number): void {
    this.tNext = tNext;
  }

  public getTCurr(): number {
    return this.tCurr;
  }

  public setTCurr(tCurr: number): void {
    this.tCurr = tCurr;
  }

  public getState(): number {
    return this.state;
  }

  public setState(state: number): void {
    this.state = state;
  }

  public getName(): string {
    return this.name;
  }

  public printInfo(): void {
    console.log(
      `${
        this.name
      } state = ${this.getState()} quantity = ${this.getQuantity()} tnext = ${this.getTNext()}`
    );
  }

  public printResult(): void {
    console.log(`${this.name} quantity = ${this.getQuantity()}`);
  }

  public getId(): number {
    return this.id;
  }

  public doStatistics(delta: number): void {}

  public setDistribution(distribution: DistributionEnum): void {
    this.distribution = distribution;
  }
}

File: lab1/pythonProject/core/core/Element.py
import random
from typing import List

from core.core.FunRand import FunRand
from core.core.Route import Route


class Element:
    next_id = 0

    def __init__(self, name, delay_mean=1.0, delay_dev=0.0):
        self.routes = []
        self.id = Element.next_id
        Element.next_id += 1
        self.name = name
        self.routing = 'by_priority'
        self.distribution = 'none'
        self.t_next = float('inf')
        self.t_curr = self.t_next
        self.delay_mean = delay_mean
        self.delay_dev = delay_dev
        self.quantity = 0
        self.state = 0

    @staticmethod
    def get_unblocked_routes(routes: List['Route'], routed_job: 'Job') -> List['Route']:
        return [route for route in routes if not route.is_blocked(routed_job)]

    @staticmethod
    def get_scaled_probabilities(routes: List['Route']) -> List[float]:
        probabilities = [route.get_probability() for route in routes]
        for i in range(1, len(probabilities)):
            probabilities[i] += probabilities[i - 1]
        total = probabilities[-1]
        return [p / total for p in probabilities]

    def set_routing(self, routing: 'Routing'):
        self.routing = routing

    def get_delay(self) -> float:
        if self.distribution == 'exponential':
            return FunRand.exponential(self.delay_mean)
        elif self.distribution == 'uniform':
            return FunRand.uniform(self.delay_mean, self.delay_dev)
        elif self.distribution == 'normal':
            return FunRand.normal(self.delay_mean, self.delay_dev)
        elif self.distribution == 'erlang':
            return FunRand.erlang(self.delay_mean, self.delay_dev)
        return self.delay_mean

    def get_delay_mean(self) -> float:
        return self.delay_mean

    def set_delay_mean(self, delay_mean: float):
        self.delay_mean = delay_mean

    def get_delay_dev(self) -> float:
        return self.delay_dev

    def set_delay_dev(self, delay_dev: float):
        self.delay_dev = delay_dev

    def get_quantity(self) -> int:
        return self.quantity

    def set_quantity(self, quantity: int):
        self.quantity = quantity

    def change_quantity(self, delta: int):
        self.quantity += delta

    def get_next_route(self, routed_job: 'Job') -> 'Route':
        if not self.routes:
            return Route(None)

        if self.routing == 'by_probability':
            return self.get_next_route_by_probability(routed_job)
        elif self.routing == 'by_priority':
            return self.get_next_route_by_priority(routed_job)
        elif self.routing == 'combined':
            return self.get_next_route_combined(routed_job)

    def get_next_route_by_probability(self, routed_job: 'Job') -> 'Route':
        unblocked_routes = self.get_unblocked_routes(self.routes, routed_job)
        if not unblocked_routes:
            return self.routes[0]

        probability = random.random()
        scaled_probabilities = self.get_scaled_probabilities(unblocked_routes)
        for i, prob in enumerate(scaled_probabilities):
            if probability < prob:
                return unblocked_routes[i]
        return unblocked_routes[-1]

    def get_next_route_by_priority(self, routed_job: 'Job') -> 'Route':
        unblocked_routes = self.get_unblocked_routes(self.routes, routed_job)
        if not unblocked_routes:
            return self.routes[0]
        return unblocked_routes[0]

    def get_next_route_combined(self, routed_job: 'Job') -> 'Route':
        selected_route = None
        for route in self.routes:
            if not route.is_blocked(routed_job):
                selected_route = route
                break

        if selected_route is None:
            return self.routes[0]

        same_priority_routes = self.find_routes_by_priority(selected_route.get_priority())
        probability = random.random()
        scaled_probabilities = self.get_scaled_probabilities(same_priority_routes)
        for i, prob in enumerate(scaled_probabilities):
            if probability < prob:
                selected_route = same_priority_routes[i]
                break
        return selected_route

    def find_routes_by_priority(self, priority: int) -> List['Route']:
        return [route for route in self.routes if route.get_priority() == priority]

    def add_routes(self, *routes: 'Route'):
        self.routes.extend(routes)
        self.routes.sort(key=lambda route: route.get_priority(), reverse=True)

    def in_act(self, job: 'Job'):
        pass

    def out_act(self):
        self.quantity += 1

    def get_tnext(self) -> float:
        return self.t_next

    def set_tnext(self, t_next: float):
        self.t_next = t_next

    def get_tcurr(self) -> float:
        return self.t_curr

    def set_tcurr(self, t_curr: float):
        self.t_curr = t_curr

    def get_state(self) -> int:
        return self.state

    def set_state(self, state: int):
        self.state = state

    def get_name(self) -> str:
        return self.name

    def print_info(self):
        print(f"{self.name} state = {self.state} quantity = {self.quantity} tnext = {self.t_next}")

    def print_result(self):
        print(f"{self.name} quantity = {self.quantity}")

    def get_id(self) -> int:
        return self.id

    def do_statistics(self, delta: float):
        pass

    def set_distribution(self, distribution: 'Distribution'):
        self.distribution = distribution